### ISSUE TITLE
Should the allocator be noexcept?

### ISSUE THREAD

**Issue by SteveBronder** [stan-dev/math#1227]
I wasn't really sure where to ask this and it's a bit out of my payscale but I figured I would file an issue.

looking at the code for the [`stack_alloc`](https://github.com/stan-dev/math/blob/develop/stan/math/memory/stack_alloc.hpp) there's two throws for `std::runtime_error` in `eight_byte_aligned_malloc` and `std::bad_alloc` in `move_to_next_block`. But do we ever had a higher scope attempt to catch and handle these?

It's kind of hard to parse online whether `noexcept` gives any serious performance improvements, though a few places online (like the article below) do explain some circumstances `noexcept` seems to be good. 

When I run the below I also see a few optimization misses caused by the if statements that are used to throw these errors.

```
g++ -std=c++1y -pthread -Wno-sign-compare   -I lib/opencl_1.2.8  -O3  -I . -I lib/eigen_3.3.3 \
 -I lib/boost_1.69.0 -I lib/sundials_4.1.0/include -I lib/gtest_1.8.1/include -I lib/gtest_1.8.1   \
-DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION \
-DSTAN_OPENCL -DOPENCL_DEVICE_ID=1 -DOPENCL_PLATFORM_ID=2 -DCL_USE_DEPRECATED_OPENCL_1_2_APIS  -fopt-info \
-DGTEST_USE_OWN_TR1_TUPLE  -c -MT test/unit/math/memory/stack_alloc_test.o \
-MM -E -MG -MP -MF test/unit/math/memory/stack_alloc_test.d test/unit/math/memory/stack_alloc_test.cpp

g++ -std=c++1y -pthread -Wno-sign-compare -I lib/opencl_1.2.8  -O3  -I . -I lib/eigen_3.3.3 -I lib/boost_1.69.0 -I lib/sundials_4.1.0/include -I lib/gtest_1.8.1/include \
-I lib/gtest_1.8.1 -I lib/gtest_1.8.1/include -I lib/gtest_1.8.1 \
-DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS \
-DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_USE_OWN_TR1_TUPLE -fopt-info-missed \
-c -o test/unit/math/memory/stack_alloc_test.o test/unit/math/memory/stack_alloc_test.cpp

```

missed optimization example
```
./stan/math/memory/stack_alloc.hpp:133:5: note: Build SLP failed: different types
```

Would making those noexcept make the allocator much more difficult to debug? 

There's a few other places the gcc compiler complains about that I think we could enable vectorization on (in particular the while loop on like 103 I think?). Is it worth thinking about this? Heuristically I figure if we call this a lot it's a good idea to try to squeeze every optimization we can out of it
https://visualstudiomagazine.com/articles/2016/10/01/noexcept.aspx

#### Current Version:
v2.19.1

**Comment by bob-carpenter** [comment]
I don't think anyone's ever thought about that in great depth. I'm not even sure what's being tested there is important.  But the thing to keep in mind is that those malloc calls happen once for a given chain on the first iteration.  So it's not really worth optimizing them unless they get in the way of some other optimization.
